app/services/concerns/Service.php

```diff
@@ -0,0 +1,18 @@
+<?php
+namespace App\services\concerns;
+
+use App\lib\PDOFactory;
+use App\lib\Request;
+
+use PDO;
+
+abstract class Service
+{
+    protected PDO $pdo;
+    
+    public function __construct(
+        protected Request $request,
+    ) {
+        $this->pdo = PDOFactory::create();
+    }
+}
```

Comment:
このクラスのコンストラクタで不要な引数を受け取っているのが気になりました。
おそらく責務としてはPDOのインスタンス化を抽象化したいことが主目的かと思います。
今の設計ですと、こちらを継承するためには個々のServiceは（使わなくても）確実にRequestクラスを受け取る必要があります。
また特定のモデルのServiceでのみ行いたいコンストラクタ処理が発生したときに、現状はここ（抽象クラス）には書きたくないのでそれぞれのServiceに書くことになり、それはそれでめんどくさそうです。

これは、
```suggestion:services/posts/PostBaseService.rb
class PostBaseService extends Service
{
    protected PDO $pdo;
    
    public function __construct(
        protected Request $request,
    ) {
        parent::__construct();
    }
}
```
のようなクラスを作ったほうが拡張性もいいかなと思いました。

Reply:
@SardineTa23 
ありがとうございます！
この部分、Request を受け取るかどうかがモデル単位で定まらないので、どういう実装にしたらいいかなと悩んでいました。
プロパティでいうと今あるパターンだけで
- request のみ
- pdo のみ
- request, pdo, id
- pdo, id

など、色々パターンがあり、リクエストに限らず pdo も余計なインスタンスが作成されてます。
パターンのそれぞれについて `ServiceWithId` のような継承元を作るとなると、どうしようかな？という感じです。
プロパティが nullable になると使用側で分岐が発生して不便なので、
```php
protected ?Request $request,
protected ?string $id,
```
のようにするのも微妙かなとは思っています 🤔 

現状、パターンそれぞれに継承元を作るのがよさそうでしょうか？

Reply:
@you-5805 まず、Modelごとなどるーるに基づいた運用にしないと破滅への道を生きそうだなというのが所感です。
その上で、現状はModelごとのものは不要と判断したら抽象クラスがRequestをコンストラクタの引数で受け取るのをやめればそれでOKだと思います！

Reply:
@SardineTa23 
> その上で、現状はModelごとのものは不要と判断したら抽象クラスがRequestをコンストラクタの引数で受け取るのをやめればそれでOKだと思います！

理解しました！ありがとうございます

Reply:
@SardineTa23  request、id は状況によって共通化できなさそうだったので、Service のコンストラクタではひとまず、pdo のインスタンス作成だけにとどめる形に修正しました。
[fix: Service properties](https://github.com/you-5805/kenshu-backend/pull/9/commits/30ec18bcc708f10dae02f30fad791d687a1f2aef)